### PR TITLE
Enable inlining into reductions

### PR DESCRIPTION
**Pull request by navahgar**
This diff enables inlining producers into reductions. It also guards against inlining reductions themselves.

Prior to this diff, if there was a reduction in the loopnest, no inlining was happening. After this change, we will inline all non-output buffers that do not correspond to a reduction.

### CODE REVIEW

File: torch/csrc/jit/tensorexpr/loopnest.cpp [hunk 1]
```diff
@@ -667,20 +668,24 @@ void LoopNest::computeInline(const Buf* b) {
     if (s->buf() == b) {
       auto reductions = NodeFinder<ReduceOp>::find(s);
       if (!reductions.empty()) {
-        throw std::logic_error("cannot inline a reduction computation");
```
Comment (nickgg) [inline torch/csrc/jit/tensorexpr/loopnest.cpp:670]: Why did you change this from throwing to returning false?

File: test/cpp/tensorexpr/test_kernel.cpp [hunk 2]
```diff
@@ -853,5 +842,88 @@ void testKernelSoftmax4D() {
   }
 }
 
+void testKernelInlineProducerIntoReduction() {
```
Comment (nickgg) [inline test/cpp/tensorexpr/test_kernel.cpp:845]: I understand adding a test in kernel because of the change there, but all the reduction tests exist in test_reduction. are you sure this is covered there also?

File: torch/csrc/jit/tensorexpr/loopnest.cpp [hunk 3]
```diff
@@ -667,20 +668,24 @@ void LoopNest::computeInline(const Buf* b) {
...
       }
       if (relevant_store != nullptr) {
-        throw std::logic_error("cannot inline Buf with multiple Tensors");
+        // Cannot inline Buf with multiple Tensors
+        return false;
       }
       relevant_store = s;
     }
   }
+  assert(relevant_store);
```
Comment (nickgg) [inline torch/csrc/jit/tensorexpr/loopnest.cpp:681]: This should be TORCH_INTERNAL_ASSERT

Comment (navahgar) [reply torch/csrc/jit/tensorexpr/loopnest.cpp]: Done.

File: test/cpp/tensorexpr/test_kernel.cpp (hunk 2 above)
Comment (navahgar) [reply test/cpp/tensorexpr/test_kernel.cpp:845]: Yes. Here are the 3 tests that cover these cases in `test_reductions.cpp`. 

Positive tests: [`testReduceInlineConsumer`](https://github.com/pytorch/pytorch/blob/master/test/cpp/tensorexpr/test_reductions.cpp#L1344), [`testReduceInlineReducerInternal`](https://github.com/pytorch/pytorch/blob/master/test/cpp/tensorexpr/test_reductions.cpp#L1398)

Negative test: [`testReduceInlineReduction`](https://github.com/pytorch/pytorch/blob/master/test/cpp/tensorexpr/test_reductions.cpp#L1311)

File: torch/csrc/jit/tensorexpr/loopnest.cpp (hunk 1 above)
Comment (navahgar) [reply torch/csrc/jit/tensorexpr/loopnest.cpp:670]: There are couple of reasons for this:
1. In order to check if a given `buf` corresponds to a reduction, we will have to do the same check in the caller before calling this function. Then, doing the check here seems unnecessary.
2. It is better to have the API `computeInline` **try** to inline the given buf and return if it was successful, rather than crashing the compiler by throwing an exception. (IMO, throwing an exception should only be for the case when the compiler is in a bad state.)
3. The logic to check whether we can inline is best kept in one location, in the function `computeInline` itself. Otherwise, we will have to do these checks in every caller location (assuming this could be called from multiple locations).

File: test/cpp/tensorexpr/test_kernel.cpp (hunk 2 above)
Comment (nickgg) [reply test/cpp/tensorexpr/test_kernel.cpp:845]: Oh write, yeah, I wrote those tests I should remember :)

File: torch/csrc/jit/tensorexpr/loopnest.cpp (hunk 1 above)
Comment (nickgg) [reply torch/csrc/jit/tensorexpr/loopnest.cpp:670]: I'm not 100% convinced, for 2 especially since we throw in other transforms and we'll have to catch above anyway.

Let's get a third opinion from the team.

**Review by nickgg — APPROVED** [review]
Consensus seems to be to remove the exceptions, so I'm happy with this.
